### PR TITLE
Replace `hess="2-point"` with `hess=SR1()` in trust-constr solver

### PR DESCRIPTION
**Pull request by MatthiasSchmidtblaicherQC**
### Problem

The `trust-constr` solver used `hess="2-point"` — a finite-difference Hessian requiring 2 × p extra objective evaluations per iteration, making it impractically slow for moderate to large feature counts.

### Fix

Two-line change in `_solvers.py`: replace `hess="2-point"` with `hess=SR1()`, a quasi-Newton approximation that reuses already-computed gradients — zero extra evaluations per iteration.

### Benchmark

Poisson GLM with monotonicity constraints. Both solvers converge to the same coefficients (max |diff| < 2e-5).

**Vary p at fixed n=500k:**

| n | p | 2-point (s) | SR1 (s) | Speedup |
|---|---|---|---|---|
| 500k | 10 | 0.87 | 0.17 | **5x** |
| 500k | 50 | 4.45 | 1.76 | **3x** |
| 500k | 100 | 42.34 | 2.25 | **19x** |

**Vary n at fixed p=50:**

| n | p | 2-point (s) | SR1 (s) | Speedup |
|---|---|---|---|---|
| 50k | 50 | 2.78 | 0.24 | **12x** |
| 500k | 50 | 4.69 | 1.77 | **3x** |

Speedup is largest when p²/n is non-negligible. At large n, the O(n·p) gradient cost dominates both methods, compressing the ratio.

### CODE REVIEW

**Comment by MatthiasSchmidtblaicherQC** [comment]
<details>
<summary>Benchmarking code</summary>

```python
import time
import warnings
from unittest.mock import patch

import numpy as np
from scipy.optimize import SR1, minimize

from glum import GeneralizedLinearRegressor


def _make_poisson_monotone_problem(n_samples, n_spline, n_other, seed=42):
    """Simulate a Poisson problem with monotonicity constraints on first n_spline cols."""
    rng = np.random.default_rng(seed)
    n_features = n_spline + n_other
    X = rng.standard_normal((n_samples, n_features))
    true_coef = np.concatenate(
        [
            np.sort(rng.uniform(0.01, 0.3, n_spline)),
            rng.uniform(-0.2, 0.2, n_other),
        ]
    )
    y = rng.poisson(np.exp(X @ true_coef + 0.5)).astype(np.float64)
    y = np.maximum(y, 0.0)

    # Monotonicity: coef[i] <= coef[i+1] for spline columns
    n_constr = n_spline - 1
    A_ineq = np.zeros((n_constr, n_features))
    for k in range(n_constr):
        A_ineq[k, k] = 1.0
        A_ineq[k, k + 1] = -1.0
    b_ineq = np.zeros(n_constr)
    return X, y, A_ineq, b_ineq



def _fit_with_hess(hess_strategy, X, y, A_ineq, b_ineq, max_iter=500, gtol=1e-8):
    """Fit a GLM using trust-constr, injecting a specific hess strategy."""

    original_minimize = minimize

    def patched_minimize(*args, **kwargs):
        kwargs["hess"] = hess_strategy
        return original_minimize(*args, **kwargs)

    family = "poisson" if A_ineq is not None else "normal"
    link = "log" if family == "poisson" else "identity"

    model = GeneralizedLinearRegressor(
        alpha=1.0,
        l1_ratio=0,
        fit_intercept=True,
        family=family,
        link=link,
        solver="trust-constr",
        max_iter=max_iter,
        gradient_tol=gtol,
        A_ineq=A_ineq,
        b_ineq=b_ineq,
    )

    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        with patch("glum._solvers.minimize", side_effect=patched_minimize):
            t0 = time.perf_counter()
            model.fit(X, y)
            elapsed = time.perf_counter() - t0

    return model.coef_, model.intercept_, elapsed, model.n_iter_


SCENARIOS = [
    # Fix n=500k (realistic insurance-size), vary p to isolate feature-count effect.
    {
        "name": "Poisson monotone (p=10)",
        "builder": lambda: _make_poisson_monotone_problem(500_000, 8, 2),
    },
    {
        "name": "Poisson monotone (p=50)",
        "builder": lambda: _make_poisson_monotone_problem(500_000, 20, 30),
    },
    {
        "name": "Poisson monotone (p=100)",
        "builder": lambda: _make_poisson_monotone_problem(500_000, 30, 70),
    },
    # Fix p=50, vary n to show sample size doesn't affect speedup.
    {
        "name": "Poisson monotone (n=50k)",
        "builder": lambda: _make_poisson_monotone_problem(50_000, 20, 30),
    },
    {
        "name": "Poisson monotone (n=500k)",
        "builder": lambda: _make_poisson_monotone_problem(500_000, 20, 30),
    },
]


def main():
    print("=" * 78)
    print("Benchmark: trust-constr Hessian — SR1() vs '2-point'")
    print("=" * 78)

    results = []

    for scenario in SCENARIOS:
        name = scenario["name"]
        X, y, A_ineq, b_ineq = scenario["builder"]()
        n_features = X.shape[1]
        print(f"\n--- {name} (n={X.shape[0]}, p={n_features}) ---")

        coef_old, icpt_old, t_old, nit_old = _fit_with_hess(
            "2-point", X, y, A_ineq, b_ineq
        )
        coef_new, icpt_new, t_new, nit_new = _fit_with_hess(SR1(), X, y, A_ineq, b_ineq)

        all_old = np.concatenate([[icpt_old], coef_old])
        all_new = np.concatenate([[icpt_new], coef_new])
        max_diff = np.max(np.abs(all_old - all_new))
        speedup = t_old / t_new if t_new > 0 else float("inf")

        print(f"  2-point:  {t_old:8.3f}s  iters={nit_old:3d}")
        print(f"  SR1():    {t_new:8.3f}s  iters={nit_new:3d}")
        print(f"  Speedup:  {speedup:.1f}x")
        print(f"  Max |coef diff|: {max_diff:.2e}")

        try:
            np.testing.assert_allclose(all_new, all_old, atol=1e-4, rtol=1e-3)
            match = "PASS"
        except AssertionError:
            match = f"WARN (max diff={max_diff:.1e})"

        results.append(
            {
                "name": name,
                "n": X.shape[0],
                "p": n_features,
                "t_old": t_old,
                "t_new": t_new,
                "speedup": speedup,
                "nit_old": nit_old,
                "nit_new": nit_new,
                "max_diff": max_diff,
                "match": match,
            }
        )

    # --- Summary table ---
    print("\n")
    print("=" * 78)
    print("SUMMARY")
    print("=" * 78)
    hdr = f"{'Scenario':<32} {'n':>7} {'p':>4} {'2-pt (s)':>9} {'SR1 (s)':>9} {'Speedup':>8} {'Coefs':>6}"
    print(hdr)
    print("-" * len(hdr))
    for r in results:
        print(
            f"{r['name']:<32} {r['n']:>7} {r['p']:>4} "
            f"{r['t_old']:>9.3f} {r['t_new']:>9.3f} "
            f"{r['speedup']:>7.1f}x {r['match']:>6}"
        )
    print()


if __name__ == "__main__":
    main()
```
</details>

File: src/glum/_solvers.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
 
 import numpy as np
 from scipy import linalg, sparse
-from scipy.optimize import LinearConstraint, fmin_l_bfgs_b, minimize
+from scipy.optimize import SR1, LinearConstraint, fmin_l_bfgs_b, minimize
```
[the remark this answers is not in the document]
Comment (MatthiasSchmidtblaicherQC) [reply src/glum/_solvers.py:10]: Wrong. upper case is first.

**Review by stanmart — APPROVED** [review]
wow 😮
